kernel-test-nohz: add logging

gratian opened the pull request:
Change the nohz_test ptest to log results, including histogram data,
to '/var/local/ptest-results/kernel-test-nohz/' to aid debugging test
failures.

This required a slight refactoring of the code that checked for test
failures to allow for logging before exiting with an error.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
kernel-test-nohz: add logging

Change the nohz_test ptest to log results, including histogram data,
to '/var/local/ptest-results/kernel-test-nohz/' to aid debugging test
failures.

This required a slight refactoring of the code that checked for test
failures to allow for logging before exiting with an error.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian commented:
Attn: @ni/rtos 

Testing:
  - built new .ipk, installed on hardware
  - verified that the log file gets created
  - verified, by hand tweaking the test thresholds that expected test failures are triggered when thresholds are exceeded.

bstreiff commented on recipes-kernel/kernel-tests/kernel-test-nohz.bb:
```diff
@@ -2,7 +2,7 @@ SUMMARY = "Linux kernel NO_HZ_FULL polling test"
 SECTION = "tests"
 LICENSE = "GPLv2"
-LIC_FILES_CHKSUM = "file://nohz_test.c;md5=d25cf78b2ec478f88868ccbc1967e9da"
+LIC_FILES_CHKSUM = "file://nohz_test.c;md5=e1d351b1ee60ef935ff16c7c0c0de455"
```
consider making this `file://nohz_test.c;beginline=1;endline=2;md5=WHATEVERTHATIS` so that the hash is only over the comment lines with the copyright/license information and doesn't need to be bumped every time.

gratian replied:
Thanks. I've added a new commit to the PR implementing this suggestion.

gratian committed:
kernel-test-nohz: tee-off and log the test output

In addition to logging the histogram data it is helpful to have a local log
of the test output too. Use `tee` to redirect a copy of the output stream
to a log file.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
kernel-test-nohz: constrain license md5 check

Constrain the license md5 check to the source lines that contain the SPDX
identifier. This eliminates the need to update the md5 sum in the recipe on
every change to the source code.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian commented:
Changes since V1:

- implemented @bstreiff 's suggestion to restrict the md5 check to the license lines in commit 270ec02446d8 ("kernel-test-nohz: constrain license md5 check");
- expanded the local logging to include the actual test output in addition to the histogram data log added by the first commit in this series. See commit 5af56e10383f ("kernel-test-nohz: tee-off and log the test output")

Additional testing:

- rebuilt recipe
- verified that the output log gets created in a test run
